Add model evaluation documentation with W&B Weave and Tables (#1558)
wandb/docs: content/en/guides/models/evaluate-models.md

---
title: Evaluate models
description:
menu:
  default:
    identifier: eval-weave
    parent: w-b-models
---

## Evaluate models with Weave

[W&B Weave](https://weave-docs.wandb.ai/) is a purpose-built toolkit for evaluating LLMs and GenAI applications. It provides comprehensive evaluation capabilities including scorers, judges, and detailed tracing to help you understand and improve model performance. Weave integrates with W&B Models, allowing you to evaluate models stored in your Model Registry.

{{< img src="/images/weave/evals.png" alt="Weave evaluation dashboard showing model performance metrics and traces" >}}

### Key features for model evaluation

* **Scorers and Judges**: Pre-built and custom evaluation metrics for accuracy, relevance, coherence, and more
* **Evaluation Datasets**: Structured test sets with ground truth for systematic evaluation
* **Model Versioning**: Track and compare different versions of your models
* **Detailed Tracing**: Debug model behavior with complete input/output traces

> * **Cost Tracking**: Monitor API costs and token usage across evaluations

Reviewer: Nit: Sentence case for the bold items :)

Author: Fixed, good call!

> ### Getting started: Evaluate a model from W&B Models Registry

Reviewer: ```suggestion
### Getting started: Evaluate a model from W&B Registry
```
There are a few instances of this phrase

> This example shows how to download a model from W&B Models Registry and evaluate it using Weave:

Reviewer: ```suggestion
Download a model from W&B Models Registry and evaluate it using Weave:
```

# Initialize Weave
weave.init("your-entity/your-project")

# Define a ChatModel that loads from W&B Registry
class ChatModel(weave.Model):
    model_name: str
    
    def model_post_init(self, __context):
        # Download model from W&B Models Registry
        run = wandb.init(project="your-project", job_type="model_download")
        artifact = run.use_artifact(self.model_name)
        self.model_path = artifact.download()
        # Initialize your model here
    
    @weave.op()
    async def predict(self, query: str) -> str:
        # Your model inference logic
        return self.model.generate(query)

# Create evaluation dataset
dataset = weave.Dataset(name="eval_dataset", rows=[
    {"input": "What is the capital of France?", "expected": "Paris"},
    {"input": "What is 2+2?", "expected": "4"},
])

# Define scorers
@weave.op()
def exact_match_scorer(expected: str, output: str) -> dict:
    return {"correct": expected.lower() == output.lower()}

# Run evaluation
model = ChatModel(model_name="wandb-entity/registry-name/model:version")
evaluation = weave.Evaluation(
    dataset=dataset,
    scorers=[exact_match_scorer]
)
results = await evaluation.evaluate(model)
```

> ### Integrating Weave evaluations with W&B Models

Reviewer: ```suggestion
### Integrate Weave evaluations with W&B Models
```

> The [Models and Weave Integration Demo](https://weave-docs.wandb.ai/reference/gen_notebooks/Models_and_Weave_Integration_Demo) shows the complete workflow for:

Reviewer: ```suggestion
The [Models and Weave integration demo notebook](https://weave-docs.wandb.ai/reference/gen_notebooks/Models_and_Weave_Integration_Demo) shows the complete workflow:
```
And change the bold below from gerunds to imperative verbs maybe?

Author: Fixed, good call :)

1. **Loading models from Registry**: Download fine-tuned models stored in W&B Models Registry
2. **Creating evaluation pipelines**: Build comprehensive evaluations with custom scorers
3. **Logging results back to W&B**: Connect evaluation metrics to your model runs
4. **Versioning evaluated models**: Save improved models back to the Registry

> Example of logging evaluation results to both Weave and W&B Models:

Reviewer: ```suggestion
Log evaluation results to both Weave and W&B Models:
```

* **Analyze errors**: Filter and query to find commonly misclassified examples and error patterns
* **Visualize rich media**: Display images, audio, text, and other media types alongside predictions and metrics

![Example of predictions table showing model outputs alongside ground truth labels](/images/data_vis/tables_sample_predictions.png)

### Basic example: Log evaluation results

```python
import wandb

# Initialize a run
run = wandb.init(project="model-evaluation")

# Create a table with evaluation results
columns = ["id", "input", "ground_truth", "prediction", "confidence", "correct"]
eval_table = wandb.Table(columns=columns)

# Add evaluation data
for idx, (input_data, label) in enumerate(test_dataset):
    prediction = model(input_data)
    confidence = prediction.max()
    predicted_class = prediction.argmax()
    
    eval_table.add_data(
        idx,
        wandb.Image(input_data),  # Log images or other media
        label,
        predicted_class,
        confidence,
        label == predicted_class
    )

# Log the table
run.log({"evaluation_results": eval_table})
```

> ### Advanced evaluation workflows

Reviewer: I'm not sure it is ideal to have two separate Advanced headings. WDYT?

Author: There is only one in each major area (Weave vs. Tables) and I think the pattern of "here's the general idea, here's the hard mode" for each of the paths makes sense? Gave them better names to highlight the approach-specific nature of those sections
